### PR TITLE
Allow config fields with defaults to be optional

### PR DESCRIPTION
**Pull request by tannerjfco**
## The Problem:
Currently, all configuration values defined in `config.yaml` are read in from file when ddev is run. This requires that all configuration values necessary for ddev to run be in the config file, without a away to run with current default values.

## The Fix:
This PR changes the behavior of config Read(). Instead of performing Unmarshal directly into `c *Config` struct, it performs Unmarshal into a new empty struct. Any non-empty values read from file are then set as their respective field values in the `c *Config` struct. This should allow for any field value that we provide a default for to be omitted from the `config.yaml` file, and still allow ddev to run using those default values.

It was additionally requested in #139 to make site name optional. This is achieved by providing a default value using the dir name the site is in. 

## The Test:
- Pick a site to test, you can use an existing valid config file or run ddev config to get a fresh one.
- This test is more fun with `DRUD_DEBUG=true` set
- Remove a previously required field from the file, such as `name`, `webimage`, `dbimage`, or `dbaimage`.
- Run ddev start. It should start successfully. If you are having fun with DEBUG set, you should also see the config values read from file, followed by the values being used for the command run.

## Automation Overview:

A test case has been added for Read()

## Related Issue Link(s):
#139 original issue report

### CODE REVIEW

File: pkg/ddevapp/config.go [hunk 1]
```diff
@@ -94,21 +98,43 @@ func (c *Config) Write() error {
...
+		return err
+	}
+
+	// Reflect non-empty values fromFile into c *Config.
+	fields, err := reflections.Items(fromFile)
 	if err != nil {
 		return err
 	}
 
+	for field, val := range fields {
```
Comment (beeradb) [inline pkg/ddevapp/config.go]: If all we're doing is setting defaults on a few fields, series of if statements checking for empty values and then setting them as appropriate seems preferable to bringing reflection into this, especially for such a limited use case?

File: pkg/ddevapp/config.go [hunk 2]
```diff
@@ -94,21 +94,41 @@ func (c *Config) Write() error {
...
+	if c.Name == "" {
+		c.Name = path.Base(c.AppRoot)
+	}
+	if c.WebImage == "" {
+		c.WebImage = version.WebImg + ":" + version.WebTag
+	}
+	if c.DBImage == "" {
+		c.DBImage = version.DBImg + ":" + version.DBTag
+	}
+	if c.DBImage == "" {
```
Comment (rfay) [inline pkg/ddevapp/config.go]: Should be DBAImage, right?

File: pkg/ddevapp/config.go [hunk 3]
```diff
@@ -94,21 +94,41 @@ func (c *Config) Write() error {
...
 	if err != nil {
 		return err
 	}
 
 	log.WithFields(log.Fields{
-		"Existing config": awsutil.Prettify(c),
+		"Read config": awsutil.Prettify(c),
 	}).Debug("Finished config read")
+
+	// If any of these values aren't defined in the config file, set them to defaults.
```
Comment (rfay) [inline pkg/ddevapp/config.go]: Thanks for the simple code here. Even though the reflection code was more elegant, this is more obvious for this simple case. We'll want reflection to simplify a really complex case for sure.

**Comment by rfay** [comment]
Thanks for this, 👍

**Comment by rfay** [comment]
Looking forward to the arrival of this one so we can remove the wired containers in the test repos.

**Pull request merged**